sphinxext/tabs.py

```diff
@@ -124,29 +143,119 @@ def build_template(tab_filter, preference):
+                tab['name'] = tab_display[index]
+                sorted_tabs[index] = tab
+
+            return filter(None, sorted_tabs)
+
+        sorter_name = 'sort' + name.title()
+        fett.Template.FILTERS[sorter_name] = sort_tabs
+
+    # Create a templated superclass for this tabs directive.
+    LegacyDirective = template.create_directive(
+        'tabs-{}'.format(name) if name else 'tabs',
+        build_template(sorter_name if tab_definitions else '', name),
         template.BUILT_IN_PATH,
         True)
 
+    class TabsDirective(LegacyDirective):
+        """A set of tabbed content. This is complex: there's four possible states we
+           handle:
+           - Legacy (YAML) Syntax
+           - Pure-RST Syntax
+           and
+           - Anonymous tabset
+           - Named tabset"""
+        required_arguments = 0
+        optional_arguments = 1
+        final_argument_whitespace = True
+        has_content = True
+        option_spec = {
+            'tabset': str,
+            'hidden': option_bool
+        }
+
+        def run(self):
+            # Transform the old YAML-based syntax into the new pure-rst syntax.
+            # This heuristic guesses whether we have the old syntax or the new.
+            # Basically, if any of the first 2 non-empty lines are "tabs:", that's a good
+            # signal that this is YAML. Why 2? One for hidden:..., one for tabs:.
+            nonempty_lines = list(line for line in self.content if line)
+            if any(line == 'tabs:' for line in nonempty_lines[:2]):
+                return LegacyDirective.run(self)
+
+            # Map the new syntax into structured data, and plug it into the old
+            # template rendering.
+            tabs_node = docutils.nodes.Element()
+            tabs_node.source, tabs_node.line = self.state_machine.get_source_and_line(self.lineno)
+            tabs_node.document = self.state.document
+            data = {
+                'hidden': self.options.get('hidden', False),
+                'tabs': []
+            }
+
+            self.state.nested_parse(
+                self.content,
+                self.content_offset,
+                tabs_node,
+                match_titles=True,
+            )
+
+            # Transform the rst nodes into structured data to plug into our template
+            for node in tabs_node.children:
+                if not isinstance(node, tab):
+                    raise self.severe(ErrorString('"tabs" may only contain "tab" directives'))
+
+                data['tabs'].append({
+                    'id': node['tabid'],
+                    'name': node['title'],
+                    'content': node['source']
+                })
+
+            return self.render(data)
+
+    return TabsDirective
+
+
+class TabInstanceDirective(docutils.parsers.rst.Directive):
+    """A single tab in a set of tabs. Contains a tabid, an optional title, and
+       a chunk of rst content."""
+    optional_arguments = 1
```

Comment:
do we need required_args here?

Reply:
Nope; the superclass initializes it to zero